### PR TITLE
Azure snitch support

### LINKED ISSUE #8593: Support Microsoft Azure snitch
Need to add support to Microsoft Azure snitch like AWS and GCE.

Can use https://github.com/yoshw/cassandra/commit/7101f3ad8f622612248a866fca37a8864f51943c for reference

### PR DESCRIPTION
**Pull request by penberg**
This add support for Azure snitch. The work is an adaptation of
AzureSnitch for Apache Cassandra by Yoshua Wakeham:

https://raw.githubusercontent.com/yoshw/cassandra/9387-trunk/src/java/org/apache/cassandra/locator/AzureSnitch.java

Also change `production_snitch_base` to protect against
a snitch implementation setting DC and rack to an empty string,
which Lubos' says can happen on Azure.

Fixes #8593

### CODE REVIEW

File: locator/azure_snitch.hh [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#pragma once
+
+#include "locator/production_snitch_base.hh"
+
+namespace locator {
+
+class azure_snitch : public production_snitch_base {
+public:
+    static constexpr auto AZURE_SERVER_ADDR = "169.254.169.254";
+    static constexpr auto AZURE_QUERY_PATH_TEMPLATE = "/metadata/instance/compute/{}?api-version=2018-04-02&format=text";
```
Comment (tarzanek) [inline locator/azure_snitch.hh:49]: why not using the latest API version?
(I don't know how quickly does older version get obsoleted and how do new or old VM have access, I'd expect Azure guys update the metadata servers frequently)
and yes, the other option is to use the version in which they introduced needed endpoints as common lowest denominator (if there is no threat of getting obsolete and unavail)

File: locator/azure_snitch.cc [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    sstring region = co_await azure_api_call(REGION_NAME_QUERY_PATH);
+    sstring azure_zone = co_await azure_api_call(ZONE_NAME_QUERY_PATH);
+
+    sstring datacenter_suffix = co_await read_property_file();
+
+    sstring azure_region = region + datacenter_suffix;
+
+    logger().info("AzureSnitch using region: {}, zone: {}.", azure_region, azure_zone);
+
+    _my_rack = azure_zone;
```
Comment (tarzanek) [inline locator/azure_snitch.cc:73]: problem is, that on some VMs I saw azure zone to be empty (while the VM had a zone)
so we have to be careful about defaults here

I didn't figure out on what does it depend, but fallback to getting the zone from somewhere (or defaulting to some at least empty string? or DEFAULT_RACK (I think we do it in some other snitch) should be good)

Comment (tarzanek) [reply locator/azure_snitch.cc:73]: OR the same goes for location
@fee-mendes do you still have some Azure VMs around?
can you do
```
curl -H Metadata:true --noproxy "*" "http://169.254.169.254/metadata/instance?api-version=2021-02-01&format=text"
```
(or the same with version 2018-04-02 )
on all your Azure VMs and share with @penberg ?

File: locator/azure_snitch.hh (hunk 1 above)
Comment (penberg) [reply locator/azure_snitch.hh:49]: @tarzanek I just converted the Java-based snitch to C++. Please let me know what the new URLs are, and I am more than happy to update the code.

**Comment by tarzanek** [comment]
re version of API, 

API versions:
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=windows#versioning

zone:
zone | Availability Zone of your virtual machine | 2017-12-01

location:
location | Azure Region the VM is running in | 2017-04-02

2020-10-01 is being rolled out
so 2020-09-01 should be everywhere by now (so it's a safe version if you ask me, but don't take me for granted)

**Comment by tarzanek** [comment]
re empty response for zone:
I think it's common that zone is empty, look here:
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=windows#instance-metadata

so I guess we should set some 
DEFAULT_RACK as zone in such case, or something similar

I have a bad feeling above was the reason why the Azure snitch isn't merged into c* yet

**Comment by tarzanek** [comment]
or support some tag with the rack name as fallback method?
I seriously don't know when and how they fill the zone in Azure, but obviously it's not a field we can rely on 100%

**Comment by yaronkaikov** [comment]
@penberg Please add a Fixme tag for https://github.com/scylladb/scylla/issues/8593

**Comment by yaronkaikov** [comment]
@penberg I think you forgot to update `dist/common/scripts/scylla_util.py` to actually use the new snitch

**Review by vladzcloudius — APPROVED** [review]
Looks good to me.
Please, address Lubos'es comment however.

**Comment by yaronkaikov** [comment]
@penberg testing your changes on Azure instance. So the good news is that scylla-server is up and running. But
when this is what i get when i check `nodetool status`
```
scyllaadm@testing-azure-snitch:~$ nodetool status
Datacenter: <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Bad Request</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Bad Request</h2>
<hr><p>HTTP Error 400. The request is badly formed.</p>
</BODY></HTML>

===================================================================================================================================================================================================================================================================================================================================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address   Load       Tokens       Owns    Host ID                               Rack
UN  10.6.0.6  280.19 KB  256          ?       c38a1b43-fe27-4999-9a59-eabf1167aefc  <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Bad Request</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Bad Request</h2>
<hr><p>HTTP Error 400. The request is badly formed.</p>
</BODY></HTML>


Note: Non-system keyspaces don't have the same replication settings, effective ownership information is meaningless
```

**Comment by yaronkaikov** [comment]
@penberg I have added today the Azure option to BYO, so once you do some changes and want me to test them, just re-build https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/1253/ (make sure you are using scylla-pkg repo and next branch since i am merged the changes today)

**Comment by penberg** [comment]
@yaronkaikov This is great news! I now just need to figure out why the HTTP request is incorrect!

**Comment by penberg** [comment]
@yaronkaikov Found the issue, and fixed it.

**Comment by penberg** [comment]
@vladzcloudius @tarzanek I would like to fix the "empty zone" issue in a generic way, and change `production_snitch_base` as follows:

```patch
diff --git a/locator/production_snitch_base.cc b/locator/production_snitch_base.cc
index 625798c50..e3b8d81a2 100644
--- a/locator/production_snitch_base.cc
+++ b/locator/production_snitch_base.cc
@@ -123,11 +123,21 @@ sstring production_snitch_base::get_endpoint_info(inet_address endpoint, gms::ap
 }

 void production_snitch_base::set_my_dc(const sstring& new_dc) {
-    _my_dc = new_dc;
+    if (!new_dc.empty()) {
+        _my_dc = new_dc;
+    } else {
+        _my_dc = default_dc;
+        logger().warn("{} snitch attempted to set DC to an empty string, falling back to {}.", get_name(), default_dc);
+    }
 }

 void production_snitch_base::set_my_rack(const sstring& new_rack) {
-    _my_rack = new_rack;
+    if (!new_rack.empty()) {
+        _my_rack = new_rack;
+    } else {
+        _my_rack = default_rack;
+        logger().warn("{} snitch attempted to set rack to an empty string, falling back to {}.", get_name(), default_rack);
+    }
 }

 void production_snitch_base::set_prefer_local(bool prefer_local) {
```

Does this seem like a reasonable solution for field engineering?

**Comment by tarzanek** [comment]
lgtm @penberg , thanks for fixing this!

**Comment by yaronkaikov** [comment]
@penberg working now !!!
```
The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.


   _____            _ _       _____  ____  
  / ____|          | | |     |  __ \|  _ \ 
 | (___   ___ _   _| | | __ _| |  | | |_) |
  \___ \ / __| | | | | |/ _` | |  | |  _ < 
  ____) | (__| |_| | | | (_| | |__| | |_) |
 |_____/ \___|\__, |_|_|\__,_|_____/|____/ 
               __/ |                       
              |___/                        

Version:
       4.6.dev-0.20210728.cef8e497278
Nodetool:
	nodetool help
CQL Shell:
	cqlsh
More documentation available at: 
	http://www.scylladb.com/doc/
By default, Scylla sends certain information about this node to a data collection server. For information, see http://www.scylladb.com/privacy/

    ScyllaDB is active.

$ nodetool status

Datacenter: eastus
==================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address   Load       Tokens       Owns    Host ID                               Rack
UN  10.6.0.8  280.08 KB  256          ?       e49d692a-c611-4e39-95c3-e405c543f007  1

Note: Non-system keyspaces don't have the same replication settings, effective ownership information is meaningless
```

**Review by yaronkaikov — APPROVED** [review]
LGTM

**Comment by penberg** [comment]
@nyh This was approved by @tarzanek and @vladzcloudius, and tested on Azure by @yaronkaikov.

**Comment by nyh** [comment]
@penberg sorry about the delay. Committed now.

**Pull request merged**